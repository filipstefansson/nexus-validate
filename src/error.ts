import { GetGen } from 'nexus/dist/core';
import { ValidationError } from 'yup';

export interface ValidatePluginErrorConfig {
  error: Error | ValidationError;
  args: any;
  ctx: GetGen<'context'>;
}

export class UserInputError extends Error {
  extensions: {
    invalidArgs: string[];
    code: string;
  };

  constructor(
    message: string,
    extensions: {
      invalidArgs: string[];
      code?: string;
    }
  ) {
    super(message);
    this.extensions = {
      ...extensions,
      code: extensions.code || 'BAD_USER_INPUT',
    };
  }
}

export const defaultFormatError = ({
  error,
}: ValidatePluginErrorConfig): Error => {
  if (error instanceof ValidationError) {
    return new UserInputError(error.message, {
      invalidArgs: error.path ? [error.path] : [],
    });
  }

  return error;
};

export { ValidationError };
