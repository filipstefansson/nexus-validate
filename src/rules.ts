import { string, number, boolean, date, object, array } from 'yup';
import { ObjectShape } from 'yup/lib/object';

export const rules = {
  string,
  number,
  boolean,
  date,
  object,
  array,
};

export type ValidationRules = typeof rules;
export { ObjectShape };
