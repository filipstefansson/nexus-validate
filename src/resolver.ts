import {
  ArgsValue,
  CreateFieldResolverInfo,
  GetGen,
  MaybePromise,
  MiddlewareFn,
} from 'nexus/dist/core';

import { ObjectShape, rules, ValidationRules } from './rules';
import { ValidatePluginConfig } from './index';
import { defaultFormatError } from './error';

export type ValidateResolver<
  TypeName extends string,
  FieldName extends string
> = (
  rules: ValidationRules,
  args: ArgsValue<TypeName, FieldName>,
  ctx: GetGen<'context'>
) => MaybePromise<ObjectShape | void>;

export const resolver = (validateConfig: ValidatePluginConfig = {}) => (
  config: CreateFieldResolverInfo
): MiddlewareFn | undefined => {
  const { formatError = defaultFormatError } = validateConfig;

  const validate: ValidateResolver<any, any> =
    config.fieldConfig.extensions?.nexus?.config.validate;

  // if the field doesn't have an validate field,
  // don't worry about wrapping the resolver
  if (validate == null) {
    return;
  }

  // if it does have this field, but it's not a function,
  // it's wrong - let's provide a warning
  if (typeof validate !== 'function') {
    console.error(
      '\x1b[33m%s\x1b[0m',
      `The validate property provided to [${
        config.fieldConfig.name
      }] with type [${
        config.fieldConfig.type
      }]. Should be a function, saw [${typeof validate}]`
    );
    return;
  }

  const args = config?.fieldConfig?.args ?? {};
  if (Object.keys(args).length === 0) {
    console.error(
      '\x1b[33m%s\x1b[0m',
      `[${config.parentTypeConfig.name}.${config.fieldConfig.name}] does not have any arguments, but a validate function was passed`
    );
  }

  return async (root, args, ctx, info, next) => {
    try {
      const schemaBase = await validate(rules, args, ctx);
      if (typeof schemaBase !== 'undefined') {
        const schema = rules.object().shape(schemaBase);
        await schema.validate(args);
      }
      return next(root, args, ctx, info);
    } catch (error) {
      throw formatError({ error, args, ctx });
    }
  };
};
