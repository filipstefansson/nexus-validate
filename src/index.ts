import { plugin } from 'nexus';
import { printedGenTyping, printedGenTypingImport } from 'nexus/dist/core';

import { ValidatePluginErrorConfig, ValidationError } from './error';
import { resolver } from './resolver';

const ValidateResolverImport = printedGenTypingImport({
  module: 'nexus-validate',
  bindings: ['ValidateResolver'],
});

const fieldDefTypes = printedGenTyping({
  optional: true,
  name: 'validate',
  description: 'Validate mutation arguments.',
  type: 'ValidateResolver<TypeName, FieldName>',
  imports: [ValidateResolverImport],
});

export interface ValidatePluginConfig {
  formatError?: (config: ValidatePluginErrorConfig) => Error;
}

export const validatePlugin = (validateConfig: ValidatePluginConfig = {}) => {
  return plugin({
    name: 'NexusValidate',
    description: 'The validate plugin provides validation for arguments.',
    fieldDefTypes: fieldDefTypes,
    onCreateFieldResolver: resolver(validateConfig),
  });
};

export * from './resolver';
export * from './error';
export { ValidationError };
