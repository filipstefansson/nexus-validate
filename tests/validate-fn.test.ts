import { graphql } from 'graphql';
import { makeSchema, objectType } from 'nexus';
import { intArg, mutationField, stringArg } from 'nexus/dist/core';
import { UserInputError, validatePlugin } from '../src/index';

describe('validatePlugin', () => {
  const consoleWarnSpy = jest
    .spyOn(console, 'warn')
    .mockImplementation(() => {});
  const consoleErrorSpy = jest
    .spyOn(console, 'error')
    .mockImplementation(() => {});

  afterEach(() => {
    jest.resetAllMocks();
  });

  const schemaTypes = [
    objectType({
      name: 'User',
      definition(t) {
        t.int('id');
      },
    }),
    mutationField('validate', {
      type: 'User',
      args: {
        email: stringArg(),
        id: intArg(),
      },
      // @ts-ignore
      validate: ({ string }, args, ctx) => {
        if (args.id !== ctx.user.id) {
          throw new UserInputError('invalid id', {
            invalidArgs: ['id'],
          });
        }
        return {
          email: string().email(),
        };
      },
      resolve: () => ({ id: 1 }),
    }),
  ];
  const testSchema = makeSchema({
    outputs: false,
    types: schemaTypes,
    nonNullDefaults: {
      output: true,
    },
    plugins: [validatePlugin()],
  });
  const mockCtx = { user: { id: 1 } };
  const testOperation = (mutation: string, schema = testSchema) => {
    return graphql(
      schema,
      `
        mutation {
          ${mutation} {
            id
          }
        }
      `,
      {},
      mockCtx
    );
  };

  it('returns error of validation error', async () => {
    const { data, errors = [] } = await testOperation(
      'validate(email: "bad@email", id: 1)'
    );
    expect(data).toBeNull();
    expect(errors.length).toEqual(1);
    expect(errors[0].message).toEqual('email must be a valid email');
    expect(errors[0].extensions).toEqual({
      code: 'BAD_USER_INPUT',
      invalidArgs: ['email'],
    });
  });

  it('can returns data if validation passed', async () => {
    const { data, errors = [] } = await testOperation(
      'validate(email: "god@email.com", id: 1)'
    );
    expect(errors).toEqual([]);
    expect(data?.validate).toEqual({ id: 1 });
  });

  it('can check args agains context', async () => {
    const { data, errors = [] } = await testOperation(
      'validate(email: "good@email.com", id: 2)'
    );
    expect(data).toBeNull();
    expect(errors.length).toEqual(1);
    expect(errors[0].message).toEqual('invalid id');
    expect(errors[0].extensions).toEqual({
      code: 'BAD_USER_INPUT',
      invalidArgs: ['id'],
    });
  });

  it('warns if validate is added in an invalid way', async () => {
    const schema = makeSchema({
      outputs: false,
      nonNullDefaults: {
        output: true,
      },
      plugins: [validatePlugin()],
      types: [
        objectType({
          name: 'ShouldWarn',
          definition(t) {
            t.int('id', {
              // @ts-ignore
              validate: () => {},
            });
          },
        }),
        mutationField('shouldWarn', {
          type: 'ShouldWarn',
          // @ts-ignore
          validate: {},
          resolve: () => ({ id: 1 }),
        }),
      ],
    });
    const { data } = await testOperation('shouldWarn', schema);
    expect(data?.shouldWarn).toEqual({ id: 1 });
    expect(consoleWarnSpy).toHaveBeenCalledTimes(1);
    expect(consoleWarnSpy.mock.calls[0]).toMatchSnapshot();
    expect(consoleErrorSpy).toHaveBeenCalledTimes(1);
    expect(consoleErrorSpy.mock.calls[0]).toMatchSnapshot();
  });
});
