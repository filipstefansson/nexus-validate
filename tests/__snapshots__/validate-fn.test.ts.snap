// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validatePlugin warns if field are missing arguments 1`] = `
Array [
  "[33m%s[0m",
  "[ShouldWarn.id] does not have any arguments, but a validate function was passed",
]
`;

exports[`validatePlugin warns if validate is not a function 1`] = `
Array [
  "[33m%s[0m",
  "The validate property provided to [shouldWarn] with type [ShouldWarn!]. Should be a function, saw [object]",
]
`;
